homeassistant/components/mqtt/mixins.py

```diff
@@ -513,12 +516,165 @@ async def cleanup_device_registry(hass, device_id, config_entry_id):
+        )
+
+    async def async_discovery_update(
+        self,
+        discovery_payload: DiscoveryInfoType | None,
+    ) -> None:
+        """Handle discovery update."""
+        discovery_hash = get_discovery_hash(self._discovery_data)
+        _LOGGER.info(
+            "Got update for %s with hash: %s '%s'",
+            self.log_name,
+            discovery_hash,
+            discovery_payload,
+        )
+        if (
+            discovery_payload
+            and discovery_payload != self._discovery_data[ATTR_DISCOVERY_PAYLOAD]
+        ):
+            _LOGGER.info(
+                "%s %s updating",
+                self.log_name,
+                discovery_hash,
+            )
+            await self.async_update(discovery_payload)
+        if not discovery_payload:
+            # Unregister and clean up the current discovery instance
+            terminate_discovery(self.hass, self._discovery_data, self._remove_signal)
+            await self._async_tear_down()
+            send_discovery_done(self.hass, self._discovery_data)
+            _LOGGER.info(
+                "%s %s has been removed",
+                self.log_name,
+                discovery_hash,
+            )
+        else:
+            # Normal update without change
+            send_discovery_done(self.hass, self._discovery_data)
+            _LOGGER.info(
+                "%s %s no changes",
+                self.log_name,
+                discovery_hash,
+            )
+            return
+
+    async def _async_device_removed(self, event) -> None:
+        """Handle the manual removal of a device."""
+        if self._skip_device_removal or not async_removed_from_device(
+            self.hass, event, self._device_id, self._config_entry_id
+        ):
+            return
+        # Avoid loop of device registry updates
+        self._skip_device_removal = True
+        # Unregister and clean up and publish an empty payload
+        # so the service is not rediscovered after a restart
+        terminate_discovery(self.hass, self._discovery_data, self._remove_signal)
+        await self._async_tear_down()
+        await async_remove_discovery_payload(self.hass, self._discovery_data)
+
+    async def _async_tear_down(self) -> None:
+        """Handle the cleanup of the discovery service."""
+        # Cleanup platform resources
+        await self.async_tear_down()
+        # remove the service for auto discovery updates and clean up the device registry
+        if not self._skip_device_removal:
+            self._skip_device_removal = True
+            await cleanup_device_registry(
+                self.hass, self._device_id, self._config_entry_id
+            )
+
+    @callback
+    def _entry_unload(self, *_: Any) -> None:
+        """Handle the unload of the config entry."""
+        terminate_discovery(self.hass, self._discovery_data, self._remove_signal)
+        # cleanup platform resources
+        self.hass.async_create_task(self._async_tear_down())
```

Comment:
This is not nice. We should, in a separate PR, teach `ConfigEntry.async_on_unload` to handle either awaitables or callbacks by making the `_on_unload` a list off `HassJob`s

See for example `homeassistant/helpers/start.py::async_at_start`

Reply:
Sure, it was added after a comment of Martin, but not reviewed after. I'll remove the added code and tests here.